File: cpp/src/AST/AST.cpp
#include "AST.h"

#include "ASTPrinter.h"

namespace sysyf {
namespace ast {
std::ostream &operator<<(std::ostream &out, const BinaryOp op) {
    switch (op) {
        case BinaryOp::OR: out << "||"; break;
        case BinaryOp::AND: out << "&&"; break;
        case BinaryOp::LT: out << "<"; break;
        case BinaryOp::LTE: out << "<="; break;
        case BinaryOp::GT: out << ">"; break;
        case BinaryOp::GTE: out << ">="; break;
        case BinaryOp::EQ: out << "=="; break;
        case BinaryOp::NEQ: out << "!="; break;
        case BinaryOp::ADD: out << "+"; break;
        case BinaryOp::SUB: out << "-"; break;
        case BinaryOp::MUL: out << "*"; break;
        case BinaryOp::DIV: out << "/"; break;
        case BinaryOp::MOD: out << "%"; break;
    }
    return out;
}
std::ostream &operator<<(std::ostream &out, const UnaryOp op) {
    switch (op) {
        case UnaryOp::POS: out << "+"; break;
        case UnaryOp::NEG: out << "-"; break;
        case UnaryOp::NOT: out << "!"; break;
    }
    return out;
}
std::ostream &operator<<(std::ostream &out, const Type op) {
    switch (op) {
        case Type::INT: out << "int"; break;
        case Type::FLOAT: out << "float"; break;
        case Type::VOID: out << "void"; break;
    }
    return out;
}

std::ostream &operator<<(std::ostream &out, const BinaryExpr &expr) { return out << ASTPrinter{}.visit(&expr); }
std::ostream &operator<<(std::ostream &out, const UnaryExpr &expr) { return out << ASTPrinter{}.visit(&expr); }
std::ostream &operator<<(std::ostream &out, const LVal &expr) { return out << ASTPrinter{}.visit(&expr); }
std::ostream &operator<<(std::ostream &out, const FuncCallExpr &expr) { return out << ASTPrinter{}.visit(&expr); }

std::ostream &operator<<(std::ostream &out, const AssignStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const ExprStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const EmptyStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const VarDefStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const IfStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const WhileStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const BreakStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const ContinueStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const ReturnStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }
std::ostream &operator<<(std::ostream &out, const BlockStmt &stmt) { return out << ASTPrinter{}.visit(&stmt); }

}  // namespace ast
}  // namespace sysyf

File: python/src/sysyf_ast/ast.py
from dataclasses import dataclass
from enum import Enum
from typing import Optional, TypeAlias


class BinaryOp(Enum):
    LT = 0
    LTE = 1
    GT = 2
    GTE = 3
    EQ = 4
    NEQ = 5
    AND = 6
    OR = 7
    ADD = 8
    SUB = 9
    MUL = 10
    DIV = 11
    MOD = 12


class UnaryOp(Enum):
    POS = 0
    NEG = 1
    NOT = 2


class Type(Enum):
    INT = 0
    FLOAT = 1
    VOID = 2


@dataclass
class BinaryExpr:
    op: BinaryOp
    lhs: "Expr"
    rhs: "Expr"


@dataclass
class UnaryExpr:
    op: UnaryOp
    exp: "Expr"


@dataclass
class LVal:
    name: str
    array_idx: Optional["Expr"]


@dataclass
class FuncCallExpr:
    func_name: str
    params: list["Expr"]


Expr: TypeAlias = BinaryExpr | UnaryExpr | LVal | FuncCallExpr | int | float


@dataclass
class ReturnStmt:
    ret: Expr | None


@dataclass
class BreakStmt:
    pass


@dataclass
class ContinueStmt:
    pass


@dataclass
class WhileStmt:
    cond: Expr
    body: "Stmt"


@dataclass
class IfStmt:
    cond: Expr
    if_body: "Stmt"
    else_body: Optional["Stmt"]


@dataclass
class ExpStmt:
    exp: Expr


@dataclass
class EmptyStmt:
    pass


@dataclass
class AssignStmt:
    var: LVal
    val: Expr


@dataclass
class VarDefStmt:
    is_const: bool
    type: Type
    name: str
    array_length: int | None
    init_vals: list[Expr] | None


BlockStmt: TypeAlias = list["Stmt"]
Stmt: TypeAlias = (
    ReturnStmt
    | BreakStmt
    | ContinueStmt
    | WhileStmt
    | IfStmt
    | BlockStmt
    | ExpStmt
    | EmptyStmt
    | VarDefStmt
    | AssignStmt
)


@dataclass
class FuncParam:
    name: str
    type: Type
    is_array: bool


@dataclass
class FuncDef:
    ret_type: Type
    name: str
    params: list[FuncParam]
    body: BlockStmt


GlobalDef: TypeAlias = FuncDef | VarDefStmt

Assembly: TypeAlias = list[GlobalDef]
